### ISSUE TITLE
Feature: Spell Queue System

### ISSUE THREAD

**Issue by lingxDEV** [azerothcore/azerothcore-wotlk#9300]
### Describe your feature request or suggestion in detail

On retail servers you are able to queue spells, meaning you can press a spell 0.1-0.4 seconds before a finishing current cast / global and the spell will be cast after the current cast.

It was implemented it on the Preservation Project (4.3.4): https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/commit/4430863a1e9a787a3a702b58adaf4c7bb639fa29

I guess it would have to be implemented differently in 3.3.5 since 4.3.4 supports it.

### Additional context

Spell queue system on Warmane: https://www.youtube.com/watch?v=lLIypI3P3mY

**Comment by Nyeriah** [comment]
Not sure how would this be implemented without breaking anything since we don't even have that opcode https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/commit/4430863a1e9a787a3a702b58adaf4c7bb639fa29#diff-d4f5deb80bb2d190160370c721b6c35de8f753e307193da6a2ef3cc539936098R229

**Comment by Kitzunu** [comment]
"**As of patch 4.1 spell casts can now be queued up 400ms before a ongoing cast finishes** or the global cooldown expires to deliver a smooth transitions between casts though emulators do not really benefit from it.
"

Going to close, as it is not blizzlike for 3.3.5

**Comment by T1ti** [comment]
This is wrong, there was a small queue system in blizzlike 3.3.5
It worked differently and was much shorter but it was there

> This used to be an in-game option, starting in Cataclysm, in the form of a slider named "Custom Lag Tolerance". It was unfortunately removed in Legion because it was deemed "too confusing".
> 
> **Spell Queuing is a feature that allows several spells to be queued up on the server - so that they can be executed immediately with no delay, regardless of how high your latency is. This is a great feature introduced in patch 2.3.0 (TBC)** and put an end to all casters needing to use /stopcasting macros and cast-bar latency indicators to perform properly. An unfortunate side-effect is that if you are a character with abilities that are cast very fast back-to-back, sometimes your character will perform attacks that you didn't want to.
> 
> As an example, Feral Druids with high haste, Heroism, and so on can sometimes cast 2x Shred instead of 1x Shred, even though you changed your mind and started mashing Ferocious Bite instead. This can be pretty annoying and can mess with your rotation. This is true for many other specs as well.
> 
> In Cataclysm, a slider was added in the game options to allow players to customize this timing window, and they recommended you put the slider to slightly above your actual latency. This worked well for some, and not for others. More specifically, it worked well for melee, but not for casters, or if adjusted higher, the opposite was true. It was hard to find a setting that worked well for both. The setting was (and still is) global, not per-character, making it annoying to set every time you re-log to a different character. This is why it was removed, presumably.
> 
> This WeakAura resolves all these problems by dynamically adjusting this spell queue window each time you log in OR change specs, to a value of your choice. See the aura Custom Options.

**Comment by T1ti** [comment]
https://github.com/Evolvee/Spell-Delay-Research/issues/1#issuecomment-443430545

**Comment by lingxDEV** [comment]
So can we reopen this request?

**Comment by lingxDEV** [comment]
@Kitzunu

**Comment by lingxDEV** [comment]
Old threat about the system: http://tobolds.blogspot.com/2010/10/wow-patch-punishes-button-mashing.html

**Comment by kurgerjkae** [comment]
This would be a great optional quality of life change. Not being blizzlike should not matter.

**Comment by heyitsbench** [comment]
> Not being blizzlike should not matter.

For a core that aims for blizzlike, it *does* matter. I'm not personally sure if it's blizzlike or not, but that *is* reason enough for it not to be implemented.

**Comment by lingxDEV** [comment]
It is Blizzlike and probably the biggest difference between private servers and retail.

**Comment by llllDaturallll** [comment]
Hello.

If we searched on mmo-champions forum, we can find some thread that evoke this feature before  the prelaunch of Cataclysm in 2010
For exemple here https://www.mmo-champion.com/threads/694955-Cast-times-near-1-0-latency-spell-queueing?highlight=spell+queue
And more, if we go the second link given at the bottom (Elitist jerks forum was an authority for TC in vanilla/TBC), we see post dated from 2009
https://web.archive.org/web/20090508120925/http://elitistjerks.com/f73/t37800-wrathcalcs_moonkin_dps_spreadsheet/p4/

And last but not least, like T1ti quote above, this feature was introduced in 2.3.0
"Client spell cast requests are now sent to the server even if your player is already casting another spell. This eliminates the need for /stopcasting in macros to compensate for latency."
https://wowpedia.fandom.com/wiki/Patch_2.3.0

Voilà

PS: On a more personal note, this is the lack of this feature that make me leave private server after sometimes
And some years later, I was curious to see if that changed and I found this github request.
